File: 85-maximal-rectangle/maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int maxiAns=0;
        int n=matrix.size();
        int m=matrix[0].size();
        vector<int>height(m,0);
        for(int i=0;i<n;i++){
           for(int j=0;j<m;j++){
             if(matrix[i][j]=='1'){
                height[j]++;
             }
             else{
                height[j]=0;
             }
           }
           maxiAns=max(maxiAns,helper(height));
        }
        return maxiAns;
    }
    int helper(vector<int>&height){
        int n=height.size();
        stack<int>s;
        int left[n];
        int right[n];
        for(int i=0;i<n;i++){
            while(!s.empty()&& height[s.top()]>=height[i]){
                s.pop();
            }
            if(s.empty()){
                left[i]=-1;
            }
            else{
                left[i]=s.top();
            }
            s.push(i);
        }
        while(!s.empty()){
            s.pop();
        }
         for(int i=n-1;i>=0;i--){
            while(!s.empty()&& height[s.top()]>=height[i]){
                s.pop();
            }
            if(s.empty()){
                right[i]=n;
            }
            else{
                right[i]=s.top();
            }
            s.push(i);
        }
        int ans=0;
        for(int i=0;i<n;i++){
            int area=height[i]*(right[i]-left[i]-1);
            ans=max(area,ans);
        }
        return ans;
    }
};

File: 85-maximal-rectangle/maximal-rectangle.java
import java.util.*;

class Solution {

    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0) return 0;

        int n = matrix.length;
        int m = matrix[0].length;

        int[] height = new int[m];
        int maxAns = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == '1') {
                    height[j]++;
                } else {
                    height[j] = 0;
                }
            }
            maxAns = Math.max(maxAns, helper(height));
        }
        return maxAns;
    }

    private int helper(int[] height) {
        int n = height.length;
        Stack<Integer> stack = new Stack<>();

        int[] left = new int[n];
        int[] right = new int[n];

        // Previous Smaller Element
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && height[stack.peek()] >= height[i]) {
                stack.pop();
            }
            left[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }

        stack.clear();

        // Next Smaller Element
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && height[stack.peek()] >= height[i]) {
                stack.pop();
            }
            right[i] = stack.isEmpty() ? n : stack.peek();
            stack.push(i);
        }

        int maxArea = 0;
        for (int i = 0; i < n; i++) {
            int area = height[i] * (right[i] - left[i] - 1);
            maxArea = Math.max(maxArea, area);
        }

        return maxArea;
    }
}
